grr/server/grr_response_server/gui/ui/components/timestamp/relative_timestamp_pipe.ts

```diff
@@ -0,0 +1,54 @@
+import {Pipe, PipeTransform} from '@angular/core';
+import {DateTime} from 'luxon';
+
+@Pipe({
+  name: 'relativeTimestamp'
+})
+export class RelativeTimestampPipe implements PipeTransform {
+  readonly oneDay = 86400000; // One day in millis
+  readonly oneHour = 3600000; // One hour in millis
+  readonly oneMin = 60000; // One minute in millis
+  readonly oneSecond = 1000; // One second in millis
+
+  transform(jsDate: Date, absoluteOnly: boolean): any {
+    if (jsDate === undefined) {
+      return "-";
+    }
+
+    const date = DateTime.fromJSDate(jsDate);
+
+    if (absoluteOnly) {
+      return this.getAbsoluteTimestamp(date);
+    }
+
+    const now = DateTime.local();
+    const timeElapsed = now.diff(date, "milliseconds").as("milliseconds");
+
+    if (timeElapsed < this.oneMin / 2) {
+      return "Just now";
+    }
+
+    if (timeElapsed < this.oneMin) {
+      return Math.floor(timeElapsed / this.oneSecond) + " seconds ago";
+    }
+
+    if (timeElapsed < this.oneHour) {
+      return Math.floor(timeElapsed / this.oneMin) + "min ago";
+    }
+
+    if (timeElapsed < this.oneDay) {
+      return Math.floor(timeElapsed / this.oneHour) + "h" +
+        Math.floor((timeElapsed % this.oneHour) / this.oneMin) + "min ago";
+    }
+
+    if (now.startOf("day").diff(date, "milliseconds").as("milliseconds") < this.oneDay) {
```

Comment:
This is fairly complicated and could lead to displaying "yesterday at xx" when in fact the `date` was the day before yesterday in the user's local timezone. While relative times ("xx hours ago") can be calculated independently of the users local timezone, an absolute wall clock time like "yesterday at xx:xx" is in the user's local timezone. The only way the be sure if `date` is yesterday in the user's timezone, is to get the user's current datetime (in their timezone), then subtract 1 day and verify if the date part is equal.

Reply:
This implementation uses local timezone of the user, because Luxon uses local timezone by default, so it shouldn't be a problem. Will change to make it more readable.